contracts/src/spaces/facets/roles/RolesBase.sol

```diff
@@ -318,6 +324,118 @@ abstract contract RolesBase is IRolesBase {
+
+    return
+      RolesStorage
+      .layout()
+      .permissionByChannelIdByRoleId[roleId][channelId].values();
+  }
+
+  function _createCustomChannelPermissions(
+    uint256 roleId,
+    bytes32 channelId,
+    string[] memory permissions
+  ) internal {
+    ChannelService.checkChannelExists(channelId);
+
+    // check role exists
+    _checkRoleExists(roleId);
+
+    RolesStorage.Layout storage rs = RolesStorage.layout();
+
+    rs.channelsByRole[roleId].add(channelId);
+
+    // check permissions is not empty
+    uint256 permissionsLen = permissions.length;
+    for (uint256 i = 0; i < permissionsLen; i++) {
+      _checkEmptyString(permissions[i]);
+      if (rs.roleById[roleId].permissions.contains(permissions[i])) {
+        revert Roles__PermissionAlreadyExists();
+      }
+      rs.permissionByChannelIdByRoleId[roleId][channelId].add(permissions[i]);
+    }
+
+    emit PermissionsAddedToChannelRole(msg.sender, roleId, channelId);
+  }
+
+  function _updateCustomChannelPermissions(
+    uint256 roleId,
+    bytes32 channelId,
+    string[] memory permissions
+  ) internal {
+    // check role exists
+    _checkRoleExists(roleId);
+
+    // check channel exists
+    ChannelService.checkChannelExists(channelId);
+
+    RolesStorage.Layout storage rs = RolesStorage.layout();
+
+    // update old permissions with the new ones
+    string[] memory currentPermissions = rs
+    .permissionByChannelIdByRoleId[roleId][channelId].values();
+    uint256 currentPermissionsLen = currentPermissions.length;
+
+    for (uint256 i = 0; i < currentPermissionsLen; i++) {
+      rs.permissionByChannelIdByRoleId[roleId][channelId].remove(
+        currentPermissions[i]
+      );
+    }
+
+    uint256 permissionsLen = permissions.length;
+    for (uint256 i = 0; i < permissionsLen; i++) {
+      _checkEmptyString(permissions[i]);
+      if (rs.roleById[roleId].permissions.contains(permissions[i])) {
+        revert Roles__PermissionAlreadyExists();
+      }
+      rs.permissionByChannelIdByRoleId[roleId][channelId].add(permissions[i]);
+    }
+
+    emit PermissionsUpdatedForChannelRole(msg.sender, roleId, channelId);
+  }
+
+  function _removeCustomChannelPermissions(
+    uint256 roleId,
+    bytes32 channelId,
+    string[] memory permissions
+  ) internal {
+    // check role exists
+    _checkRoleExists(roleId);
+
+    // check channel exists
+    ChannelService.checkChannelExists(channelId);
+
+    RolesStorage.Layout storage rs = RolesStorage.layout();
+    StringSet.Set storage permissionsSet = rs.permissionByChannelIdByRoleId[
+      roleId
+    ][channelId];
+
+    // get current permissions
+    uint256 permissionsLen = permissions.length;
+
+    for (uint256 i = 0; i < permissionsLen; i++) {
+      _checkEmptyString(permissions[i]);
+      if (!permissionsSet.contains(permissions[i])) {
+        revert Roles__PermissionDoesNotExist();
+      }
+      permissionsSet.remove(permissions[i]);
+    }
```

Comment:
If the create method above modifies roles like so,

```
rs.channelsByRole[roleId].add(channelId);
```

Would we also potentially want to do the remove here?

Reply:
it's a set, not an array.